src/main/java/io/pravega/leaderelection/LeaderElection.java

```diff
@@ -0,0 +1,493 @@
+        }
+    }
+
+    @RequiredArgsConstructor
+    private static class DeclareDead extends HeartbeatUpdate {
+        private static final long serialVersionUID = 1L;
+        private final String name;
+
+        @Override
+        public LiveInstances applyTo(LiveInstances state, Revision newRevision) {
+            Map<String, InstanceInfo> tempInstances = new HashMap<>(state.liveInstances);
+            tempInstances.remove(name);
+            return new LiveInstances(state.scopedStreamName,
+                    newRevision,
+                    Collections.unmodifiableMap(tempInstances),
+                    state.vectorTime,
+                    state.leaderName);
+        }
+    }
+
+    @RequiredArgsConstructor
+    private static class SetLeader extends HeartbeatUpdate {
+        private static final long serialVersionUID = 1L;
+        private final String newLeader;
+        @Override
+        public LiveInstances applyTo(LiveInstances state, Revision newRevision) {
+                Map<String, InstanceInfo> tempInstances = new HashMap<>(state.liveInstances);
+
+                for(String key: tempInstances.keySet()) {
+                    InstanceInfo info = tempInstances.get(key);
+                    tempInstances.put(key, new InstanceInfo(info.timestamp, 1, info.timeout));
+                }
+
+                return new LiveInstances(state.scopedStreamName,
+                        newRevision,
+                        Collections.unmodifiableMap(state.liveInstances),
+                        state.vectorTime,
+                        newLeader);
+        }
+    }
+
+    @RequiredArgsConstructor
+    private static class AddMember extends HeartbeatUpdate {
+        private static final long serialVersionUID = 1L;
+        private final String name;
+        private final long time;
+
+        @Override
+        public LiveInstances applyTo(LiveInstances state, Revision newRevision) {
+            Map<String, InstanceInfo> tempInstances = new HashMap<>(state.liveInstances);
+            tempInstances.put(name, new InstanceInfo(time, 1, INITIAL_TIMEOUT));
+            return new LiveInstances(state.scopedStreamName,
+                    newRevision,
+                    Collections.unmodifiableMap(tempInstances),
+                    state.vectorTime,
+                    state.leaderName);
+        }
+    }
+
+    /**
+     * Send new heartbeat to Pravega through the state Synchronizer.
+     * It will do three things:
+     * 1. update timestamp and timeout.
+     * 2. remove all dead instances.
+     * 3. If leader dead or not exist, select a new one.
+     */
+    private void sendHeartbeat() {
+        stateSync.updateState((state, updates) -> {
+            long vectorTime = state.getVectorTime() + 1;
+            InstanceInfo instanceInfo = state.liveInstances.get(instanceId);
+            long newTimes = instanceInfo.times + 1;
+            double newTimeout = (instanceInfo.timeout * instanceInfo.times +
```

Comment:
Here and below: Feel free to take advantage of a larger line width. In Pravega we use 120.